### ISSUE TITLE
Possible unknown race condition

### ISSUE THREAD

**Issue by fennewald** [elves/elvish#1312]
Was writing a quick hacky script. I had 9 ssh hosts, wanted to check which are up. Wrote the following snippet:
```
range 1 10 | peach [n]{
    try {
        ssh -o ConnectTimeout=5 syncing-$n exit 2>/dev/null
        echo $n 'is up'
    } except err {
        echo $n 'is' (styled 'not' inverse) 'up'
    }
} | sort
```

I sort at the end so that it outputs in nice order but doesn't bother testing the ssh serially. I've ran this script about a hundred times, and exactly once it produced this output:
```
1 is up is up
2 is not up
3 is up
48 is up
5 is not up
6 is not up
7 is not up
9 is not up
```
Every other time it produced a totally sensible output. I'm pretty confident it was not user interaction that caused the behavior. If this isn't enough for an issue, I understand, I will close, just very unsure of what happened.

**Comment by fennewald** [comment]
```
1 is up is up
2 is not up
3 is up
48 is up
5 is not up
6 is not up
7 is not up
9 is not up
```
I'm sorry, didn't post the full output. My mistake.

**Comment by fennewald** [comment]
Another misfire to report:
```
 is not up
1 is up
2 is not up
3 is up
4 is up
6 is not up
75 is not up
8 is up
9 is not up
```

**Comment by krader1961** [comment]
TL;DR: Should the builtin `echo`, and related commands, write their output atomically regardless of the number of arguments?

@fennewald: That was a good problem report, but does illustrate a misunderstanding of byte stream output. The behavior you noted is easy to produce in most other languages. Sticking to Elvish this is a reproducer that does not involve `ssh`:

```
#!/usr/bin/env elvish
range 1 10 | peach [n]{
    try {
        elvish -c 'exit (if (!= 0 (% $args[0] (randint 1 10))) { put 0 } else { put 1 })' $n
        echo $n 'is up'
    } except err {
        echo $n 'is' (styled 'not' inverse) 'up'
    }
}
```

The above program results in the undesired output almost as often as it produces the desired output. This variant never produces unexpected output:

```
#!/usr/bin/env elvish
range 1 10 | peach [n]{
    try {
        elvish -c 'exit (if (!= 0 (% $args[0] (randint 1 10))) { put 0 } else { put 1 })' $n
        echo $n' is up'
    } except err {
        echo $n' is'(styled 'not' inverse)'up'
    }
}
```

The only difference in the two programs is the number of arguments passed to `echo` (more than one in the first example and one in the second). Should `echo` implicitly concatenate all arguments before writing them to the output byte stream? That could be expensive. It also does not solve the fundamental problem posed by this issue.

**Comment by fennewald** [comment]
@krader1961 -- you're right, I did misunderstand the byte stream. Thank you both for the explanation and the example!

In a slight tangent and some mildly cruel irony, I originally wrote the snippet with one argument, and only separated it out for aesthetics.

As far as the implementation of echo, I worry that differing behavior depending on argument count is very unintuitive. As I mentioned, I originally wrote the snippet with only one argument for the echo. When I switched to the multi-argument style, I reasoned that echo's behavior shouldn't change based on argument count. I know I was wrong, but I feel like I was being reasonable when I assumed that, and I worry that future elves will make the same assumption.

**Comment by krader1961** [comment]
> As far as the implementation of echo, I worry that differing behavior depending on argument count is very unintuitive.

The problem, as I alluded to in my previous comment, is that changing `echo` to implicitly concatenate its arguments before issuing a single write doesn't solve the fundamental problem. Consider if you had instead written your program like this to indicate that a ssh was in progress:

```
print $n ...
try {
    ssh
    echo is up
} except {
    echo is down
}
```

And it isn't just the `echo` command. It's every command, including external programs you might run, which write to the stdout byte stream. Changing the behavior of `echo` "fixes" just one way in which you might get unexpected interleaved output.

**Comment by fennewald** [comment]
I understand now. I am going to close this issue then, because I think that is definitely beyond scope of my original question. I think this merits a lot more discussion, but probably not here.
